### PR TITLE
Include zscale

### LINKED ISSUE #27: IRAF z-scale
Everyone's favorite DS9/IRAF feature!

This feature would allow some way of defining the image scale according to IRAF's z-scale. This method requires some simple statistics and fitting and should return a minimum and maximum color. 

## Example Usage
```julia
using AstroImages
img = load("image.fits")
cmin, cmax = zscale(img, args...)

plot(img, cmin=cmin, cmax=cmax)
```

## References
- [DS9 source code](https://github.com/SAOImageDS9/SAOImageDS9/blob/bcf8233a7ce1c81a0fa287d824d9f0ec62ed2e18/ds9/library/scale.tcl#L644-L693)
- [A hard to parse written description](https://iraf.net/forum/viewtopic.php?showtopic=134139)
- [astropy](https://docs.astropy.org/en/stable/api/astropy.visualization.ZScaleInterval.html)

### PR DESCRIPTION
**Pull request by Devagio**
This DS9/IRAF feature allows defining the image scale according to IRAF's z-scale and returns a minimum and maximum color. This fixes #27

### CODE REVIEW

File: src/zscale.jl [hunk 1]
```diff
@@ -0,0 +1,248 @@
+MAX_REJECT = 0.5
+MIN_NPIXELS = 5
+GOOD_PIXEL = 0
+BAD_PIXEL = 1
+KREJ = 2.5
+MAX_ITERATIONS = 5
```
Comment (giordano) [inline src/zscale.jl:6]: Please, have a read of the performance tips in the Julia manual: https://docs.julialang.org/en/v1/manual/performance-tips/.  **Not** using non-constant global variables is the very first issue.  Some of these variables aren't even used in more than a function, so making them global isn't even useful.

File: src/zscale.jl [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    result = Array{Float64}(undef, n_pxls)
+    
+    k = 1
+    for i in 1:rows, j in 1:cols
+        result[k] = image[i, j]
+        k += 1
+    end
+
+    return result
+end
```
Comment (giordano) [inline src/zscale.jl:26]: This is what `image[:]` does, this function isn't needed

File: src/zscale.jl [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    # Figure out which pixels to use for the zscale algorithm
+    # Returns the 1-d array samples
+    # Sample in a square grid, and return the first maxpix in the sample
+
+    nc = size(image)[1]
+    nl = size(image)[2]
+
+    stride = maximum([1.0, sqrt((nc - 1) * (nl - 1) / maxpix)])
+    stride = Int128(round(stride))
```
Comment (giordano) [inline src/zscale.jl:39]: Why `Int128`?

File: src/zscale.jl [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+function zsc_sample(image, maxpix)
+
+    # Figure out which pixels to use for the zscale algorithm
+    # Returns the 1-d array samples
+    # Sample in a square grid, and return the first maxpix in the sample
+
+    nc = size(image)[1]
+    nl = size(image)[2]
+
+    stride = maximum([1.0, sqrt((nc - 1) * (nl - 1) / maxpix)])
```
Comment (giordano) [inline src/zscale.jl:38]:
```suggestion
    stride = max(1.0, sqrt((nc - 1) * (nl - 1) / maxpix))
```

File: src/zscale.jl [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+function zsc_compute_sigma(flat, badpix)
+
+    # Compute the rms deviation from the mean of a flattened array.
+    # Ignore rejected pixels
+
+    # Accumulate sum and sum of squares
+    npix = length(badpix)
+    sumz = 0
+    sumsq = 0
```
Comment (giordano) [inline src/zscale.jl:55]: These variables will likely be type-unstable

File: src/zscale.jl [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    for i in 1:rows, j in 1:cols
+        result[k] = image[i, j]
+        k += 1
+    end
+
+    return result
+end
+
+
+function zsc_sample(image, maxpix)
```
Comment (giordano) [inline src/zscale.jl:29]: What is this function actually doing and how?  Please add docstrings to functions.

File: src/zscale.jl [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    stride = maximum([1.0, sqrt((nc - 1) * (nl - 1) / maxpix)])
+    stride = Int128(round(stride))
+
+    samples = flatten(image[1:stride:end, 1:stride:end])
+    
+    return samples[1:maxpix]
+end
+
+
+function zsc_compute_sigma(flat, badpix)
```
Comment (giordano) [inline src/zscale.jl:47]: Also here (and everywhere) a docstring would be useful.  Please also explain the meaning of the `mean` or `sigma` equal to `nothing` (which I'm not sure is a good idea)

File: src/zscale.jl [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    end
+
+    upper = maximum([n1, n2])
+    return final[1:upper]
+end
+
+
+function zsc_fit_line(samples, npix, krej, ngrow, maxiter)
+
+    # calculate slope and intercept for the algorithm
```
Comment (giordano) [inline src/zscale.jl:109]: Which algorithm?  There is no reference to what you're doing here.

File: src/zscale.jl [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        Scaling factor for determining min and max. Larger values increase the
+        difference between min and max values used for display.
+
+    Returns
+    -------
+    (z1, z2)
+    =#
+
+    # Sample the image
+    nsamples = minimum([nsamples, length(image)])
```
Comment (giordano) [inline src/zscale.jl:209]:
```suggestion
    nsamples = min(nsamples, length(image))
```

File: src/zscale.jl [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+    
+    zmin = samples[1]
+    zmax = samples[end]
+
+    # For a one-indexed array
+    if npix % 2 == 1
+        center_pixel = Int128(round((npix + 1) / 2))
+        median = samples[center_pixel]
+    else
+        Int128(round(npix / 2))
```
Comment (giordano) [inline src/zscale.jl:222]: What is this line doing?

**Review by giordano — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution!  Besides a few comments I wrote down (only the first time I found a specific issue,  but they're repeated several times across this file), however there are many things to improve:

1. there are no docstrings and no references for what you're doing, it's really hard to follow what you do if you don't explain it
2. there are no tests anywhere
3. this code isn't actually used anywhere

**Comment by Devagio** [comment]
Okay thank you for such a detailed response; I'll begin working on your suggestions. 
Apologies for the mistakes and sub-optimal code, I'm new to both Julia and open-source.

File: src/zscale.jl (hunk 10 above)
Comment (Devagio) [reply src/zscale.jl:222]: It is supposed to be:
`center_pixel = Int128(round(npix / 2))`
Will rectify it with other changes.

**Comment by mileslucas** [comment]
@Devagio I've coded zscale in Julia over at [PlotUtils.jl](https://github.com/JuliaPlots/PlotUtils.jl/blob/master/src/intervals.jl), you can use that as a reference if you like. It's been a while since I wrote that code, you may find ways to improve upon it. Otherwise you may wish to just copy it, this decouples any reliance on the Plots.jl-verse.

**Comment by giordano** [comment]
Given the fact we're going towards #29, I'm not sure we really need to add a zscale implementation here.

**Comment by sefffal** [comment]
With regards to #29, right now I have it re-exporting zscale from PlotUtils so that one can pass `clims=zscale ∘ collect`. 
The `collect` is necessary for now with either this PR or the one from PlotUtils, since for performance reasons, functions passed to `clims` should accept an iterable. This way they can loop through the data without allocations when possible. 
I think the z-scale algorithm does need a full materialized copy of the array, so perhaps it's type signature could be to loosened to allow iterables and then just `collect` them.

**Comment by icweaver** [comment]
Howdy all, just wanted to check back in to see if this PR is still needed now?

**Comment by cgarling** [comment]
It probably isn't worth duplicating the implementation from PlotUtils.jl. If that function is re-exported here, as was mentioned above, that seems like a good solution.
